app/scripts/lib/crypto/recovery-keys.js

```diff
@@ -0,0 +1,164 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * This file contains utilities to help create, bundle and unbundle encrypted
+ * recovery key data.
+ *
+ * For more encryption details, check out
+ + https://github.com/mozilla/fxa-auth-server/blob/master/docs/recovery_keys.md
+ */
+const Base32 = require('./base32');
+
+function importFxaCryptoDeriver() {
+  return import(/* webpackChunkName: "fxaCryptoDeriver" */ 'fxaCryptoDeriver');
+}
+
+function required(object, name) {
+  if (! object) {
+    throw new Error(`Missing ${name}`);
+  }
+}
+
+const RECOVERY_KEY_VERSION = '1';
+const ENCRYPTION_ALGORITHM = 'A256GCM';
+const HKDF_SHA_256 = 'HKDF-SHA-256';
+
+function getRecoveryKeyVersion() {
+  return RECOVERY_KEY_VERSION;
+}
+
+module.exports = {
+  /**
+   * Generate a random base32 recovery key. The recovery key
+   * string is prepended with version information.
+   *
+   * @param {Integer} length Length of string to generate
+   * @returns {Promise<string>} recovery key
+   */
+  generateRecoveryKey: function (length) {
+    length = length || 32;
+    return Base32.generate(length - 1)
+      .then((key) => {
+        return getRecoveryKeyVersion() + key;
+      });
+  },
+
+  /**
+   * Get the current version of the recovery keys.
+   *
+   * @returns {string} current recovery key version
+   */
+  getCurrentRecoveryKeyVersion: function () {
+    return getRecoveryKeyVersion();
+  },
+
+  /**
+   * Retrieves the recovery JWK from the uid and recovery key. The
+   * recoveryJwk can be used to encrypt and decrypt data.
+   *
+   * @param {String} uid Uid of user
+   * @param {String} recoveryKey Recovery key
+   * @returns {Promise} A promise that will be fulfilled with JWK
+   */
+  getRecoveryJwk: function (uid, recoveryKey) {
+    return importFxaCryptoDeriver().then(({jose}) => {
+      required(uid, 'uid');
+      required(recoveryKey, 'recoveryKey');
+
+      return Base32.decode(recoveryKey)
+        .then((keyMaterial) => {
+          keyMaterial = Buffer.from(keyMaterial, 'utf8');
```

Comment:
Yep, I updated my python script to see if I could match your test vectors here, and `expectedRecoveryKeyId = 'c3cdf70fbf0789046156ac68fc3db8fa'` indicates that you're passing the hex-encoded keyMaterial into HKDF rather than the raw bytes.  My tests suggest the correct value should be `'6aa248931704886f54ac64b81b111bc0'`.

Reply:
@vbudhram this still says "utf8" as second argument.  I think it's now ignored because `keyMaterial` is already raw bytes, but you should probably remove it to avoid confusion.